File: ITG Hiring Challenge Front-end/src/services/ProductService.js
import axiosInstance from "@/config/axiosInstance";

class ProductService {
  getProducts(size, page, keyword, minPrice, maxPrice, sortDirection) {
    const params = {
      keyword,
      size,
      page,
      minPrice,
      maxPrice,
      sortDirection,
    };
    const query = Object.entries(params)
      .filter(([key, value]) => value !== "")
      .map(([key, value]) => `${key}=${value}`)
      .join("&");

    const url = `/products?${query}`;
    return axiosInstance.get(url);
  }

  getProductsByCategoryId(
    categoryId,
    size,
    page,
    keyword,
    minPrice,
    maxPrice,
    sortDirection
  ) {
    const params = {
      categoryId,
      size,
      page,
      keyword,
      minPrice,
      maxPrice,
      sortDirection,
    };

    const query = Object.entries(params)
      .filter(([key, value]) => value !== "")
      .map(([key, value]) => `${key}=${value}`)
      .join("&");

    const url = `/products/findByCategoryId?${query}`;
    return axiosInstance.get(url);
  }

  getProductsByBrandId(
    brandId,
    size,
    page,
    keyword,
    minPrice,
    maxPrice,
    sortDirection
  ) {
    const params = {
      brandId,
      size,
      page,
      keyword,
      minPrice,
      maxPrice,
      sortDirection,
    };

    const query = Object.entries(params)
      .filter(([key, value]) => value !== "")
      .map(([key, value]) => `${key}=${value}`)
      .join("&");

    const url = `/products/findByBrandId?${query}`;
    return axiosInstance.get(url);
  }

  getProductById(productId) {
    return axiosInstance.get(`/products/findById?id=${productId}`);
  }

  getTopFiveProduct() {
    return axiosInstance.get("/products/findTopFiveProduct");
  }

  getNewProduct() {
    return axiosInstance.get("/products/findNewProduct");
  }

  getNewOrderProduct() {
    return axiosInstance.get("/products/findNewOrderFiveProduct");
  }
}

const productService = new ProductService();
export default productService;

File: ITG Hiring Challenge Backend/automotive/src/main/java/com/biricik/automotive/business/abstracts/ProductService.java
package com.biricik.automotive.business.abstracts;
import java.math.BigDecimal;
import java.util.List;
import java.util.Optional;

import org.springframework.web.multipart.MultipartFile;

import com.biricik.automotive.business.requests.productRequest.AddImageForProductRequest;
import com.biricik.automotive.business.requests.productRequest.CreateProductRequest;
import com.biricik.automotive.business.responses.PaginatedGenericResponse;
import com.biricik.automotive.business.responses.productResponses.AddImageForProductResponse;
import com.biricik.automotive.business.responses.productResponses.CreateProductReponse;
import com.biricik.automotive.business.responses.productResponses.FindByBrandIdProductResponse;
import com.biricik.automotive.business.responses.productResponses.FindByCategoryIdProductResponse;
import com.biricik.automotive.business.responses.productResponses.FindByIdProductResponse;
import com.biricik.automotive.business.responses.productResponses.FindNewProductResponse;
import com.biricik.automotive.business.responses.productResponses.FindTopFiveProductResponse;
import com.biricik.automotive.business.responses.productResponses.GetAllProductResponse;


public interface ProductService {

	public CreateProductReponse addProduct(CreateProductRequest createProductRequest,MultipartFile file);
	
	
	public AddImageForProductResponse addImageForProduct(AddImageForProductRequest addImageForProductRequest,MultipartFile file);
	
	public  PaginatedGenericResponse<GetAllProductResponse> getAll(String keyword,int size,int page , BigDecimal minPrice, BigDecimal maxPrice,  Optional<String> sortDirection);
	
	public PaginatedGenericResponse<FindByBrandIdProductResponse> findByBrandId(String keyword,int brandId,int size,int page , 
			BigDecimal minPrice, BigDecimal maxPrice, Optional<String> sortDirection);
	
	public FindByIdProductResponse findById(int productId);


	public PaginatedGenericResponse<FindByCategoryIdProductResponse> findByCategoryId(String keyword, int categoryId,
		int size,int page, BigDecimal minPrice, BigDecimal maxPrice,  Optional<String> sortDirection); 
	
	
	public void changeProductQuantity(int shoppingCartId);
	
	
	
	public List<FindTopFiveProductResponse> findTopFiveProduct();


	public List<FindNewProductResponse> findNewProduct();
	
	public List<FindNewProductResponse> findNewOrderFiveProduct();
	

}
